backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java

```diff
@@ -0,0 +1,237 @@
+                        activityLogConfig.getPresentationExchangeStatesForTasks(), true));
+            }
+        }
+        return results;
+    }
+
+    public List<ActivityItem> getItems(ActivitySearchParameters parameters) {
+        List<ActivityItem> results = new ArrayList<>();
+        results.addAll(getActivityListItems(parameters));
+        results.addAll(getTaskListItems(parameters));
+        results.sort(Comparator.comparingLong(ActivityItem::getUpdatedAt).reversed());
+        return results;
+    }
+
+    private List<ActivityItem> getConnectionRequests(@Nullable ActivityType type, List<ConnectionState> states,
+            Boolean incoming) {
+        List<ActivityItem> results = new ArrayList<>();
+        if (type == null || type == ActivityType.CONNECTION_REQUEST) {
+            Iterable<Partner> partners = partnerRepo.findByStateIn(states);
+            for (Partner p : partners) {
+                if (incoming) {
+                    // then we are looking for tasks...
+                    if (p.getIncoming() != null) {
+                        results.add(getConnectionRequestItem(p, true));
+                    }
+                } else {
+                    results.add(getConnectionRequestItem(p, false));
+                }
+            }
+        }
+        return results;
+    }
+
+    private List<ActivityItem> getPresentationExchanges(ActivityType type, List<PresentationExchangeState> states,
+            Boolean task) {
+        List<ActivityItem> results = new ArrayList<>();
+        if (type == null || type == ActivityType.PRESENTATION_EXCHANGE) {
+            Iterable<PartnerProof> proofs = proofRepository.findByStateIn(states);
+            for (PartnerProof p : proofs) {
+                results.add(getPresentationExchangeItem(p, task));
+            }
+        }
+        return results;
+    }
+
+    private ActivityItem getConnectionRequestItem(Partner p, Boolean task) {
+        ActivityRole role = (p.getIncoming() == null) ? ActivityRole.CONNECTION_REQUEST_SENDER
+                : ActivityRole.CONNECTION_REQUEST_RECIPIENT;
+        ActivityType type = ActivityType.CONNECTION_REQUEST;
+
+        ActivityState state;
+        switch (p.getState()) {
+        case ACTIVE:
+        case RESPONSE:
+            state = ActivityState.CONNECTION_REQUEST_ACCEPTED;
+            break;
+        default:
+            switch (role) {
+            case CONNECTION_REQUEST_SENDER:
+                state = ActivityState.CONNECTION_REQUEST_SENT;
+                break;
+            default:
+                state = ActivityState.CONNECTION_REQUEST_RECEIVED;
+            }
+        }
+
+        Long updatedAt = p.getUpdatedAt().toEpochMilli();
+        String linkId = p.getId().toString();
+        return ActivityItem.builder()
+                .role(role)
+                .state(state)
+                .type(type)
+                .partner(partnerManager.refreshPartner(p.getId()).get())
```

Comment:
See comment below. refresh takes a long long time to complete, when all you need is to load the partner and convert it to the api model.